lucet-wiggle/tests/wasi.rs

```diff
@@ -0,0 +1,456 @@
+        unimplemented!("proc_exit")
+    }
+
+    fn proc_raise(&self, _sig: types::Signal) -> Result<()> {
+        unimplemented!("proc_raise")
+    }
+
+    fn sched_yield(&self) -> Result<()> {
+        unimplemented!("sched_yield")
+    }
+
+    fn random_get(&self, _buf: GuestPtr<u8>, _buf_len: types::Size) -> Result<()> {
+        unimplemented!("random_get")
+    }
+
+    fn sock_recv(
+        &self,
+        _fd: types::Fd,
+        _ri_data: &types::IovecArray<'_>,
+        _ri_flags: types::Riflags,
+    ) -> Result<(types::Size, types::Roflags)> {
+        unimplemented!("sock_recv")
+    }
+
+    fn sock_send(
+        &self,
+        _fd: types::Fd,
+        _si_data: &types::CiovecArray<'_>,
+        _si_flags: types::Siflags,
+    ) -> Result<types::Size> {
+        unimplemented!("sock_send")
+    }
+
+    fn sock_shutdown(&self, _fd: types::Fd, _how: types::Sdflags) -> Result<()> {
+        unimplemented!("sock_shutdown")
+    }
+}
+
+/// Test the above generated code by running Wasm code that calls into it.
+#[test]
+fn main() {
+    // The `init` function ensures that all of the host call functions are linked into the
+    // executable.
+    crate::init();
+
+    // Temporary directory for outputs.
+    let workdir = TempDir::new().expect("create working directory");
+
+    // Build a C file into a Wasm module. Use the wasi-sdk compiler, but do not use the wasi libc
+    // or the ordinary start files,
+    // which will together expect various aspects of the Wasi trait to actually work. The C file
+    // only imports one function, `args_sizes_get`.
+    let wasm_build = Link::new(&["tests/wasi_guest.c"])
+        .with_cflag("-nostartfiles")
+        .with_link_opt(LinkOpt::NoDefaultEntryPoint)
+        .with_link_opt(LinkOpt::AllowUndefinedAll)
+        .with_link_opt(LinkOpt::ExportAll);
+    let wasm_file = workdir.path().join("out.wasm");
+    wasm_build.link(wasm_file.clone()).expect("link wasm");
+
+    // We used lucet_wiggle to define the hostcall functions, so we must use it to define our
+    // bindings as well. This is a good thing! No more bindings json files to keep in sync with
+    // implementations.
+    let witx_doc = witx::load(&["../wasi/phases/snapshot/witx/wasi_snapshot_preview1.witx"])
+        .expect("load snapshot 1 witx");
+    let bindings = lucet_wiggle_generate::bindings(&witx_doc);
+
+    // Build a shared object with Lucetc:
+    let native_build = Lucetc::new(wasm_file).with_bindings(bindings);
+    let so_file = workdir.path().join("out.so");
+    native_build
+        .shared_object_file(so_file.clone())
+        .expect("build so");
+
+    // Load shared object into this executable.
+    let module = DlModule::load(so_file).expect("load so");
+
+    // Create an instance:
+    let region = MmapRegion::create(1, &Limits::default()).expect("create region");
+    let mut inst = region.new_instance(module).expect("create instance");
```

Comment:
I _hugely_ appreciate that you've documented the code in this test with thorough comments. Especially for a test like this, that involves a fair amount of setup. Nice!